cpp/src/interpolate/cubic_spline.cu

```diff
@@ -0,0 +1,404 @@
+ * limitations under the License.
+ */
+
+#include <cuspatial/cubic_spline.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/filling.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include "cusparse.h"
+#include <cuspatial/utility.hpp>
+
+namespace { // anonymous
+
+struct parallel_search {
+  template <typename T>
+  std::enable_if_t<std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& curve_ids,
+                  cudf::column_view const& prefixes,
+                  cudf::column_view const& query_coords,
+                  cudf::mutable_column_view const& result,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      const T* T_ = t.data<T>();
+      const int32_t* CURVE_IDS_ = curve_ids.data<int32_t>();
+      const int32_t* PREFIXES_ = prefixes.data<int32_t>();
+      const T* QUERY_COORDS_ = query_coords.data<T>();
+      T* RESULT_ = result.data<T>();
+      thrust::for_each(rmm::exec_policy(stream)->on(stream),
+        thrust::make_counting_iterator<int>(0),
+        thrust::make_counting_iterator<int>(query_coords.size()),
+        [T_, CURVE_IDS_, PREFIXES_, QUERY_COORDS_, RESULT_] __device__
+        (int index) {
+          int curve = CURVE_IDS_[index];
+          int len = PREFIXES_[curve+1] - PREFIXES_[curve];
+          int h = PREFIXES_[curve];
+          int dh = PREFIXES_[curve] - (curve);
+          // O(n) search, can do log(n) easily
+          for(int32_t i = 0 ; i < len ; ++i) {
+            if((T_[h+i]+0.0001 - QUERY_COORDS_[index]) > 0.00001) {
+              RESULT_[index] = dh+i-1;
+              if(i == 0) RESULT_[index] = index-curve;
+              return;
+            }
+          }
+          // TODO: Important failure case:
+          // This will use the final set of coefficients
+          // for t_ values that are outside of the original
+          // interpolation range.
+          RESULT_[index] = h+len - 2;
+      });
+  };
+  template <typename T>
+  std::enable_if_t<not std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& curve_ids,
+                  cudf::column_view const& prefixes,
+                  cudf::column_view const& query_coords,
+                  cudf::mutable_column_view const& result,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      CUDF_FAIL("Non-floating point operation is not supported.");
+  }
+};
+
+struct interpolate {
+  template <typename T>
+  std::enable_if_t<std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& ids,
+                  cudf::column_view const& coef_indexes,
+                  cudf::table_view const& coefficients,
+                  cudf::mutable_column_view const& result,
```

Comment:
Prefer returning results. It doesn't make sense to pass a result by const&.

Reply:
I'm concerned about returning results - my result sizes is known at runtime and preallocated, which I thought would be a performance gain. Returning result from any of my functors doesn't make sense because the functor is called once for each `offset`: allocations would be expensive and more importantly there's nowhere to put the resulting vector of result columns. I think it should be preallocated to follow the SoA form, which is currently the standard in `cuspatial`.

Reply:
The interpolate functor is only called once.

Reply:
👍